### PR TITLE
#4 sht30 Driver

### LINKED ISSUE #4: Develop SHT30 Temp Sensor Driver
SHT30-DIS-B10KS <- part #

https://github.com/Northeastern-Electric-Racing/NERduino/blob/main/src/sht30.cpp
(Link to old nerduino code)

https://github.com/Sensirion/embedded-sht/tree/master/sht3x
(Sensirons implementation)

https://sensirion.com/media/documents/213E6A3B/63A5A569/Datasheet_SHT3x_DIS.pdf
(Data sheet)

### PR DESCRIPTION
**Pull request by Peyton-McKee**
Sht30 driver based off previous implemenation and sensirion implementation updated to work with i2c handling in c

closes #4

### CODE REVIEW

File: Drivers/SHT30/sht30.c [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    cmd_t cmd;
+    cmd.cmdVal = SWITCHBYTES(SHT30_READSTATUS);
+
+    SHT30write(&cmd, 2);
+    uint8_t data[3];
+    SHT30read(data, 3);
+    uint16_t status = data[0];
+    status <<= 8;
+    status |= data[1];
+    return status;
```
Comment (nwdepatie) [inline Drivers/SHT30/sht30.c:45]: I think we should utilize return-based error codes and trickle up errors, so we prob want to pass in a pointer as arguments and then fill the data

File: Drivers/SHT30/sht30.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#include <stdint.h>
+
+//SHT30-DIS Humidity Sensor
+/**
+ * https://www.mouser.com/datasheet/2/682/Sensirion_Humidity_Sensors_SHT3x_Datasheet_digital-971521.pdf  --Datasheet
+ * 
+ */
+
+#define SWITCHBYTES(x)             (((x<<8) | (x>>8)) & 0xFFFF)      //Switches the first and second byte (This is required for all command values)
```
Comment (nwdepatie) [inline Drivers/SHT30/sht30.h:16]: I think we need to either move this to the .c file or get rid of it, I bet there is a macro like this already built into C. I made this a while ago and regret it lol

File: Drivers/SHT30/sht30.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#define SHT30_START_CMD_NCS        0x2400   //Start measurement command with clock streching disabled and high repeatability 
+#define SHT30_READSTATUS           0xF32D   //Read Out of Status Register
+#define SHT30_CLEARSTATUS          0x3041   //Clear Status
+#define SHT30_SOFTRESET            0x30A2   //Soft Reset
+#define SHT30_HEATEREN             0x306D   //Heater Enable
+#define SHT30_HEATERDIS            0x3066   //Heater Disable
+
+#define SHT30_REG_HEATER_BIT       0x0d     //Status Register Heater Bit
+
+typedef struct
```
Comment (nwdepatie) [inline Drivers/SHT30/sht30.h:30]: I think we should get rid of this struct lol, it kinda serves no purpose. It also feels like it should be a union but its just a struct lmao

File: Drivers/SHT30/sht30.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{
+    uint16_t cmdVal;
+    uint8_t cmd2Bytes[2];
+} cmd_t;
+
+typedef struct
+{
+    uint16_t statusReg;
```
Comment (nwdepatie) [inline Drivers/SHT30/sht30.h:38]: We should read thru the datasheet and see what else we might want to add in this. We also need to investigate the STM32 I2C HAL and see if we need references to any internal I2C bus addresses or struct-based APIs like regmap in the Linux kernel

File: Drivers/SHT30/sht30.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+} sht30_t;
+
+sht30_t* create_SHT30();
+
+void SHT30write(cmd_t *cmd, uint8_t num_bytes);
```
Comment (nwdepatie) [inline Drivers/SHT30/sht30.h:43]: We shouldn't have this defs in the header since we don't wanna expose them. They should just be static in the C file

**Review by nwdepatie — CHANGES_REQUESTED** [review]
General comments: I think you should take a look at this PR that Hamza made a while ago, kinda similar comments. Basically I think we should use return based error handling, probably want a function to init the struct and remove the function to deallocate the memory, and leverage static functions wherever we can to avoid namespace conflicts. Also prefix everything with ```sht30_``` to indicate where the function is coming from since we don't have classes to organize

File: Drivers/SHT30/sht30.c (hunk 1 above)
Comment (Peyton-McKee) [reply Drivers/SHT30/sht30.c:45]: what type would an error be. Like a code or a string message?

File: Drivers/SHT30/sht30.c [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+static void sht30_handle_status_reg(sht30_t *sht30)
+{
+    sht30_write(sht30, SWITCHBYTES(SHT30_READSTATUS), 2);
+    uint8_t data[3];
+    SHT30read(&data, 3);
+    uint16_t status = data[0];
+    status <<= 8;
+    status |= data[1];
+    sht30->status_reg = status;
+    //IDK how to get a error from this rn lol
```
Comment (HamzaIqbal69) [inline Drivers/SHT30/sht30.c:40]: Theres an stm Hal error type that you can use to return errors if you use the hal, just look up stm32 i2c driver on youtube and it should explain it

**Review by nwdepatie — APPROVED** [review]
I made changes and now I like :)

**Comment by nwdepatie** [comment]
Thoughts on squashing? Should we combine all commits into one big commit or separate into "copy from NERduino" and then "modified driver"?

File: Drivers/NER/include/sht30.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+#define SHT30_READSTATUS           0xF32D   //Read Out of Status Register
+#define SHT30_CLEARSTATUS          0x3041   //Clear Status
+#define SHT30_SOFTRESET            0x30A2   //Soft Reset
+#define SHT30_HEATEREN             0x306D   //Heater Enable
+#define SHT30_HEATERDIS            0x3066   //Heater Disable
+
+#define SHT30_REG_HEATER_BIT       0x0d     //Status Register Heater Bit
+
+
+typedef struct
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Drivers/NER/include/sht30.h:30]: sounds good

File: Drivers/NER/src/sht30.c [hunk 8]
```diff
@@ -0,0 +1,141 @@
+#include "sht30.h"
+#include <stdio.h>
+#include <stdbool.h>
+
+#define SWITCHBYTES(x)             (((x<<8) | (x>>8)) & 0xFFFF)      //Switches the first and second byte (This is required for all command values)
+
+/* Creates a new sht30 Driver*/
+sht30_t* sht30_init(sht30_t *sht30)
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Drivers/NER/src/sht30.c]: makes sense

File: Drivers/NER/src/sht30.c [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+/**
+ * @brief Writes to the SHT30
+ * @param temp_sensor: Pointer to the sht30_t struct
+ * @param data: Pointer to the data to be written
+ * @param reg: Register to be written to
+ * @param size: the size of the data being received
+ */
+static HAL_StatusTypeDef sht30_read_reg(sht30_t *temp_sensor, uint8_t *data[], uint16_t reg, uint8_t size)
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Drivers/NER/src/sht30.c]: is that how that works?

File: Drivers/NER/src/sht30.c [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        return status;
+    }
+    uint8_t data[3];
+    status = sht30_read_reg(temp_sensor, &data, SWITCHBYTES(SHT30_READSTATUS), 3);
+    if (status != HAL_OK)
+    {
+        return status;
+    }
+    uint16_t register_status = data[0];
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Drivers/NER/src/sht30.c]: uhhh code celanliness?

File: Drivers/NER/src/sht30.c [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    return HAL_OK;
+}
+
+/**
+ * @brief Calculates the CRC by using the polynomial x^8 + x^5 + x^4 + 1
+ * @param data: the data to use to calculate the CRC
+*/
+static uint8_t calculate_crc(uint16_t data) {
+    uint8_t crc = 0xFF;
+    crc ^= (data >> 8);
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Drivers/NER/src/sht30.c:45]: I typically like one line per operation

File: Drivers/NER/src/sht30.c [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+            crc = (crc << 1);
+        }
+    }
+    return crc;
+}
+
+/* Resets the SHT30 */
+HAL_StatusTypeDef sht30_reset(sht30_t *temp_sensor)
+{
+    return sht30_write_reg(temp_sensor, SWITCHBYTES(SHT30_SOFTRESET), 0, 2);
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Drivers/NER/src/sht30.c]: i think we do actually but right now the initializing isnt actually doing anything

File: Drivers/NER/src/sht30.c [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+    union sht30_data_decoding data;
+    status = sht30_read_reg(temp_sensor, data.dataArray, SWITCHBYTES(SHT30_START_CMD_WCS), 6);
+    if (status != HAL_OK)
+    {
+        return status;
+    }
+
+    uint16_t temp = data.raw_data.temp;
+    if (data.raw_data.tempCRC != calculate_crc(temp))
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Drivers/NER/src/sht30.c]: yes

**Comment by Peyton-McKee** [comment]
> Thoughts on squashing? Should we combine all commits into one big commit or separate into "copy from NERduino" and then "modified driver"?

I never squash but if we were I think the two separate commits would be good

**Pull request merged**